pkg/controller/federatedcluster/controller.go

```diff
@@ -160,82 +149,138 @@ func (cc *ClusterController) Run(stopChan <-chan struct{}) {
+		// Increment the result run has there is no change in cluster condition
+		storedData.resultRun++
+	}
+
+	if storedData.resultRun < threshold {
+		// Success/Failure is below threshold - leave the probe state unchanged.
+		probeTime := clusterStatus.Conditions[0].LastProbeTime
+		clusterStatus = storedData.clusterStatus
+		setLastProbeTime(clusterStatus, probeTime)
+	} else {
+		if clusterStatusEqual(clusterStatus, storedData.clusterStatus) {
+			// preserve the last transition time
+			setLastTransitionTime(clusterStatus, storedData.clusterStatus.Conditions[0].LastTransitionTime)
 		}
-		clusterStatusNew := clusterClient.GetClusterHealthStatus()
-		if !statusFound {
-			glog.Infof("There is no status stored for cluster: %v before", cluster.Name)
-		} else {
-			hasTransition := false
-			if len(clusterStatusNew.Conditions) != len(clusterStatusOld.Conditions) {
-				hasTransition = true
-			} else {
-				for i := 0; i < len(clusterStatusNew.Conditions); i++ {
-					if !(strings.EqualFold(string(clusterStatusNew.Conditions[i].Type), string(clusterStatusOld.Conditions[i].Type)) &&
-						strings.EqualFold(string(clusterStatusNew.Conditions[i].Status), string(clusterStatusOld.Conditions[i].Status))) {
-						hasTransition = true
-						break
-					}
-				}
-			}
+	}
+
+	if !clusterStatusEqual(clusterStatus, storedData.clusterStatus) {
+		// Reset the result run
+		storedData.resultRun = 1
+	}
 
-			if !hasTransition {
-				for j := 0; j < len(clusterStatusNew.Conditions); j++ {
-					clusterStatusNew.Conditions[j].LastTransitionTime = clusterStatusOld.Conditions[j].LastTransitionTime
-				}
+	return clusterStatus
+}
+
+func updateClusterZoneAndRegion(clusterStatus *fedv1a1.FederatedClusterStatus, clusterSpecifics *ClusterData,
+	clusterClient *ClusterClient, clusterName string) *fedv1a1.FederatedClusterStatus {
+	zone, region, err := clusterClient.GetClusterZones()
+	if err != nil {
+		glog.Warningf("Failed to get zones and region for cluster %s: %v", clusterName, err)
+	} else {
+		// If new zone & region are empty, preserve the old ones so that user configured zone & region
+		// labels are effective
+		if clusterSpecifics.clusterStatus != nil {
+			if len(zone) == 0 {
+				zone = clusterSpecifics.clusterStatus.Zone
+			}
+			if len(region) == 0 {
+				region = clusterSpecifics.clusterStatus.Region
 			}
 		}
+		clusterStatus.Zone = zone
+		clusterStatus.Region = region
+	}
+
+	return clusterStatus
+}
 
-		if utilfeature.DefaultFeatureGate.Enabled(features.CrossClusterServiceDiscovery) {
-			zone, region, err := clusterClient.GetClusterZones()
-			if err != nil {
-				glog.Warningf("Failed to get zones and region for cluster %s: %v", cluster.Name, err)
-			} else {
-				if len(zone) == 0 {
-					zone = cluster.Status.Zone
-				}
-				if len(region) == 0 {
-					region = cluster.Status.Region
-				}
-				clusterStatusNew.Zone = zone
-				clusterStatusNew.Region = region
+func isClusterReady(clusterStatus *fedv1a1.FederatedClusterStatus) bool {
```

Comment:
Please reuse `util.IsClusterReady` instead of adding a new implementation.

Reply:
Could not use `util.IsClusterReady`, as the parameter it takes is `FederatedCluster` object against the `FederatedClusterStatus` as being used in this PR. It would be lot of unrelated change. so skipping it and hope its fine.

Reply:
Why not refactor, then, so that one method takes `FederatedCluster` and the other `FederatedClusterStatus`?  Or update callers of `IsClusterReady` to pass status instead (since passing cluster just to use status violates the law of demeter)?  I'm not clear why duplicating logic like this is ever desirable.

Reply:
Refactored and reused `util.IsClusterReady` now.